From 64238f81d1e0f8be10777253ac490ec24165ae5f Mon Sep 17 00:00:00 2001
From: Tanmay Jagdale <tanmay@marvell.com>
Date: Tue, 4 Jan 2022 20:37:19 +0530
Subject: [PATCH 03/12] soc: octeontx2: ccu: Fix waymask configuration

commit 77bd2c8c87a3f6209f57e54f78608f1cd8186fde from
git@git.assembla.com:cavium/WindRiver.linux.git

Setting the LTG way-mask bits is causing the same bits to be set in
DTG way-mask bitfields as well. This is due to improper value passed
to FIELD_SET macro during DTG way configuration. Fix this by writing
the way-mask value passed by user directly.

Change-Id: I571e40ad83cd6c67ddebee76e05948b372c26933
Signed-off-by: Tanmay Jagdale <tanmay@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/68658
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <sgoutham@marvell.com>
(cherry picked from commit 447affef51683d8a0e2a6d75480f3903fbf371dd)
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/69046
Tested-by: Sunil Kovvuri Goutham <sgoutham@marvell.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 drivers/soc/marvell/octeontx2-ccu/ccu.c | 25 ++++---------------------
 1 file changed, 4 insertions(+), 21 deletions(-)

diff --git a/drivers/soc/marvell/octeontx2-ccu/ccu.c b/drivers/soc/marvell/octeontx2-ccu/ccu.c
index 9158337bc8ed..60bfa1438791 100644
--- a/drivers/soc/marvell/octeontx2-ccu/ccu.c
+++ b/drivers/soc/marvell/octeontx2-ccu/ccu.c
@@ -17,20 +17,7 @@
 #define CCU_BASE	0x87E050000000
 #define CCS_BASE	0x87E087100000
 
-/* m - bit mask
- * y - value to be written in the bitrange
- * x - input value whose bitrange to be modified
- */
-#define FIELD_SET(m, y, x)		\
-	(((x) & ~(m)) |			\
-	FIELD_PREP((m), (y)))
-
-#define CCS_MPARX_MASK(pid)	\
-	(0x1000		|	\
-	((pid) & 0xff) << 3)
-
-#define MPARX_MASK_LTG		GENMASK_ULL(13, 0)
-#define MPARX_MASK_DTG		GENMASK_ULL(33, 14)
+#define CCS_MPARX_MASK(pid)	(0x1000 | ((pid) & 0xff) << 3)
 
 #define CCUX_TADX_MPARX_ACNT(ccu, tad, pid)	\
 	(0x401000		|		\
@@ -99,14 +86,10 @@ static inline void ccureg_write(u64 offset, u64 val)
 }
 
 /* Mask LLC ways for a partition id */
-static inline void ccsreg_mparmask_set(int mparid, u32 waymask)
+static inline void ccsreg_mparmask_set(int mparid, u64 waymask)
 {
-	u64 val;
-
-	val = ccsreg_read(CCS_MPARX_MASK(mparid));
-	val = FIELD_SET(MPARX_MASK_LTG, waymask, val);
-	val = FIELD_SET(MPARX_MASK_DTG, waymask, val);
-	ccsreg_write(CCS_MPARX_MASK(mparid), val);
+	/* Bits [13:0] mask LTG ways, bits [33:14] mask DTG ways */
+	ccsreg_write(CCS_MPARX_MASK(mparid), waymask);
 }
 
 static ssize_t otx2_ccu_config_write(struct file *file, const char *buf,
-- 
2.34.1

